q2_demux/_demux.py

```diff
@@ -504,3 +509,102 @@ def emp_paired(seqs: BarcodePairedSequenceFastqIterator,
     _write_metadata_yaml(result)
 
     return result, ec_details_fmt
+
+
+def partition_samples_single(demux: SingleLanePerSampleSingleEndFastqDirFmt,
+                             num_partitions: int = None
+                             ) -> SingleLanePerSampleSingleEndFastqDirFmt:
+    return _partition_helper(demux, num_partitions, paired=False)
+
+
+def partition_samples_paired(demux: SingleLanePerSamplePairedEndFastqDirFmt,
+                             num_partitions: int = None
+                             ) -> SingleLanePerSamplePairedEndFastqDirFmt:
+    return _partition_helper(demux, num_partitions, paired=True)
+
+
+def _partition_helper(demux, num_partitions, paired):
+    """ Deal with partitioning logic that is largely the same regardless of
+        single or paired.
+    """
+    partitioned_demux = {}
+    df = demux.manifest.view(pd.DataFrame)
+
+    # Adjust based on if we are in the single or paired end case
+    if paired:
+        result_class = SingleLanePerSamplePairedEndFastqDirFmt
+    else:
+        result_class = SingleLanePerSampleSingleEndFastqDirFmt
+
+    # Make sure we are partitioning on samples if no number of partitions or
+    # too many partitions specified and warn if they specified too many
+    # partitions
+    num_samples = df.shape[0]
+    if num_partitions is None:
+        num_partitions = num_samples
+    elif num_partitions > num_samples:
+        warnings.warn("You have requested a number of partitions"
```

Comment:
Do you think this is a better approach than just erroring if `num_partitions > num_samples`? I wonder if this will just get overlooked and cause users to assume `num_partitions` is what was used instead of an error message that forces them to change this param to something that makes sense for their data.